### DISCUSSION TITLE
R Squared Negative Value

### DISCUSSION THREAD

**Discussion by shubhambooking2** [google/meridian#167] [General]
iam using a weekly level national model and incorporating the holdout period as well. Iam a bit surprised that RSquared for the holdout period is coming extremely high on negative side (> -1) where as metrics like MAPE/WMAPE seems pretty reasonabl. Any ideas why this potentially be and how this could be fixed

**Comment by alexwakim** [comment]
It seems suspicious that wMAPE and R-squared would give such different results. The team is currently investigating whether this is a possible bug. A couple of questions for you,

- How many weeks are in your data?
- How many weeks are held out in holdout_id?
- How were the holdout weeks selected?
- Does your data exhibit really strong trend and seasonality?

Although it’s suspicious that wMAPE and R-squared give such different results, it is possible. wMAPE measures the prediction accuracy relative to the size of the actual revenue. R-squared measures the prediction accuracy relative to the variance of the actual revenue. It is possible that prediction errors could be small on the one scale, but large on the other. 

A couple of other things to consider,

- R-squared is squaring the errors whereas wMAPE is an absolute value. Thus, R-squared is more sensitive to outliers.
- If you are using multiple knots to model time effects, be mindful of how many knots are used and where the holdout observations are located relative to the knots. If all of the time points surrounding a knot are held out, then there is no data to inform the knot value. In that case, the knot's posterior distribution will be driven by the prior, which may result in poor predictions. This scenario can occur if, for example, the holdout sample is concentrated all at the beginning or end of the data.
- Prediction accuracy is calculated on the revenue scale (`kpi * revenue_per_kpi`) unless `revenue_per_kpi = None` (`kpi` is the modeled dependent variable).

The `Analyzer.expected_vs_actual_data` method may help as you consider what is described above.

**Comment by shubhambooking2** [comment]
Thanks alex for looking into this.

My training data comprises of 156 weeks worth of data points
Holdout set comprises of 13 weeks of data points
Holdout week selection strategy - immediate next 13 weeks post 156 weeks are part of holdout set
data definietly has a strong trend and seasonal component - to address this iam using a knot for each year and indeed i agree with your point here that  i have the holdout set at the end so it is very likely knot's posterior being driven by prior

**Comment by alexwakim** [comment]
If you randomly select the holdout, does it remedy the issue?

**Comment by shubhambooking2** [comment]
Indeed with use of random holdout, the Issue seems to be not present and Rsquared values are well behaved
